File: generator-common/src/main/java/io/github/potjerodekool/openapi/common/generate/AbstractGenerator.java
package io.github.potjerodekool.openapi.common.generate;

import io.github.potjerodekool.codegen.Environment;
import io.github.potjerodekool.codegen.Language;
import io.github.potjerodekool.codegen.io.Location;
import io.github.potjerodekool.codegen.model.element.ElementKind;
import io.github.potjerodekool.codegen.model.element.Modifier;
import io.github.potjerodekool.codegen.model.util.QualifiedName;
import io.github.potjerodekool.codegen.model.util.StringUtils;
import io.github.potjerodekool.codegen.template.ImportOrganiser;
import io.github.potjerodekool.codegen.template.TemplateBasedGenerator;
import io.github.potjerodekool.codegen.template.model.TCompilationUnit;
import io.github.potjerodekool.codegen.template.model.annotation.Annot;
import io.github.potjerodekool.codegen.template.model.element.TypeElem;
import io.github.potjerodekool.codegen.template.model.element.VariableElem;
import io.github.potjerodekool.codegen.template.model.expression.SimpleLiteralExpr;
import io.github.potjerodekool.codegen.template.model.type.ClassOrInterfaceTypeExpr;
import io.github.potjerodekool.codegen.template.model.type.TypeExpr;
import io.github.potjerodekool.openapi.common.ApiConfiguration;
import io.github.potjerodekool.openapi.common.GeneratorConfig;
import io.swagger.models.HttpMethod;
import io.swagger.v3.oas.models.OpenAPI;
import io.swagger.v3.oas.models.Operation;
import io.swagger.v3.oas.models.parameters.Parameter;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

import static io.github.potjerodekool.openapi.common.util.OpenApiUtils.findJsonMediaType;
import static io.github.potjerodekool.openapi.common.util.OpenApiUtils.isMultiPart;
import static io.github.potjerodekool.openapi.common.util.OpenApiUtils.isImageOrVideo;
import static io.github.potjerodekool.openapi.common.util.StringUtils.toValidClassName;

public abstract class AbstractGenerator implements OpenApiWalkerListener {

    private final String basePackageName;

    private final String modelPackageName;

    private final Language language;

    private final OpenApiTypeUtils typeUtils;

    private final Map<String, TCompilationUnit> compilationUnitMap = new HashMap<>();

    private final Environment environment;

    protected AbstractGenerator(final GeneratorConfig generatorConfig,
                                final ApiConfiguration apiConfiguration,
                                final OpenApiTypeUtils typeUtils, final Environment environment) {
        this.basePackageName = Objects.requireNonNull(
                apiConfiguration.basePackageName(),
                generatorConfig::basePackageName
        );
        this.modelPackageName = apiConfiguration.modelPackageName();
        this.language = generatorConfig.language();
        this.typeUtils = typeUtils;
        this.environment = environment;
    }

    public String getBasePackageName() {
        return basePackageName;
    }

    public String getModelPackageName() {
        return modelPackageName;
    }

    public OpenApiTypeUtils getTypeUtils() {
        return typeUtils;
    }

    public Collection<TCompilationUnit> getCompilationUnits() {
        return compilationUnitMap.values();
    }

    public void generate(final OpenAPI openAPI) {
        OpenApiWalker.walk(openAPI, this);

        final var generator = new TemplateBasedGenerator();
        final var filer = environment.getFiler();

        final var importOrganiser = new ImportOrganiser();

        compilationUnitMap.values().forEach(cu -> {
            importOrganiser.organiseImports(cu);

            final var code = generator.doGenerate(cu);
            final var clazz = cu.getElements().getFirst();

            final var resource = filer.createResource(
                    Location.SOURCE_OUTPUT,
                    cu.getPackageName(), clazz.getSimpleName() + ".java");

            resource.writeToOutputStream(code.getBytes());
        });
    }

    protected TypeElem findOrCreateClass(final String path,
                                         final Operation operation) {
        final var className = resolveClassName(path, operation);
        final var qualifiedName = QualifiedName.from(className);
        final var packageName = qualifiedName.packageName().toString();
        final var simpleName = qualifiedName.simpleName().toString();

        final var cu = this.compilationUnitMap.computeIfAbsent(className, (key) ->
                createCompilationUnit(packageName, simpleName));

        return cu.getElements().stream()
                .findFirst()
                .orElse(null);
    }

    private TCompilationUnit createCompilationUnit(final String packageName,
                                                   final String simpleName) {
        final var cu = new TCompilationUnit(language);
        cu.packageName(packageName);

        final var clazz = createClass(simpleName);
        cu.element(clazz);
        return cu;
    }

    protected TypeElem createClass(final String simpleName) {
        final var date = DateTimeFormatter.ISO_DATE_TIME.format(LocalDateTime.now());

        return new TypeElem()
                .modifier(Modifier.PUBLIC)
                .simpleName(simpleName)
                .annotation(
                        new Annot("javax.annotation.processing.Generated")
                                .value(new SimpleLiteralExpr(getClass().getName()))
                                .value("date", new SimpleLiteralExpr(date))
                );
    }

    protected String resolveClassName(final String path,
                                      final Operation operation) {
        return generateClasName(path, operation);
    }

    protected String generateClasName(final String path,
                                      final Operation operation) {
        final var className = resolveQualifiedName(path, operation);
        final var suffix = classNameSuffix();
        return suffix == null ? className : className + suffix;
    }

    protected abstract String classNameSuffix();

    private String resolveQualifiedName(final String path,
                                        final Operation operation) {
        String simpleName;

        final var pathElements = path.split("/");

        final var simpleNameBuilder = new StringBuilder();

        for (final String pathElement : pathElements) {
            if (!pathElement.startsWith("{")) {
                simpleNameBuilder.append(StringUtils.firstUpper(pathElement));
            }
        }

        simpleName = simpleNameBuilder.toString();

        simpleName = toValidClassName(simpleName);

        final String apiName;

        if (operation.getTags() == null || operation.getTags().isEmpty()) {
            apiName = StringUtils.firstUpper(simpleName);
        } else {
            final var firstTag = operation.getTags().getFirst();
            apiName = Arrays.stream(firstTag.split("-"))
                    .map(StringUtils::firstUpper)
                    .collect(Collectors.joining());
        }

        return basePackageName + "." + apiName;
    }

    protected List<VariableElem> createParameters(final OpenAPI api,
                                                  final Operation operation,
                                                  final HttpMethod httpMethod) {
        final List<VariableElem> parameters = new ArrayList<>();

        if (operation.getParameters() != null) {
            parameters.addAll(
                    operation.getParameters().stream()
                            .map(parameter -> createParameter(api, parameter))
                            .toList()
            );
        }

        final var requestBody = operation.getRequestBody();

        if (requestBody != null) {
            final var bodyMediaType = findJsonMediaType(requestBody.getContent());
            final TypeExpr bodyType;

            if (bodyMediaType != null) {
                final var resolved = SchemaResolver.resolve(api, bodyMediaType);

                if (httpMethod == HttpMethod.PATCH) {
                    final var schemaName = resolved.name();
                    final var bodyObjectSchema = resolved.schema();

                    if (schemaName.toLowerCase().contains("patch")) {
                        bodyType = typeUtils.createType(
                                api,
                                bodyObjectSchema,
                                modelPackageName,
                                ContentTypes.JSON,
                                requestBody.getRequired()
                        );
                    } else {
                        final var type = (ClassOrInterfaceTypeExpr) typeUtils.createType(
                                api,
                                bodyObjectSchema,
                                modelPackageName,
                                ContentTypes.JSON,
                                requestBody.getRequired()
                        );

                        final var simpleName = "Patch" + schemaName;
                        type.name(modelPackageName + "." + simpleName);
                        bodyType = type;
                    }
                } else {
                    bodyType = typeUtils.createType(
                            api,
                            bodyMediaType,
                            modelPackageName,
                            ContentTypes.JSON,
                            requestBody.getRequired()
                    );
                }
            } else if (
                    isMultiPart(requestBody.getContent())
                            || isImageOrVideo(requestBody.getContent())) {
                bodyType = typeUtils.createMultipartTypeExpression(api);
            } else {
                bodyType = new ClassOrInterfaceTypeExpr("java.lang.Object");
            }

            final var bodyParameter = new VariableElem()
                    .kind(ElementKind.PARAMETER)
                    .modifier(Modifier.FINAL)
                    .type(bodyType)
                    .simpleName("body");

            parameters.add(bodyParameter);
        }

        return parameters;
    }

    protected VariableElem createParameter(final OpenAPI openAPI,
                                           final Parameter openApiParameter) {
        var type = typeUtils.createType(
                openAPI,
                openApiParameter.getSchema(),
                modelPackageName,
                null,
                openApiParameter.getRequired()
        );

        if (Boolean.TRUE.equals(openApiParameter.getRequired())) {
            type = typeUtils.asNonNull(type);
        }

        return new VariableElem()
                .kind(ElementKind.PARAMETER)
                .modifier(Modifier.FINAL)
                .type(type)
                .simpleName(openApiParameter.getName());
    }
}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/generate/AbstractGenerator.kt
package io.github.potjerodekool.openapi.common.generate

import io.github.potjerodekool.codegen.Environment
import io.github.potjerodekool.codegen.io.Location
import io.github.potjerodekool.codegen.model.element.ElementKind
import io.github.potjerodekool.codegen.model.element.Modifier
import io.github.potjerodekool.codegen.model.util.QualifiedName
import io.github.potjerodekool.codegen.model.util.StringUtils
import io.github.potjerodekool.codegen.template.ImportOrganiser
import io.github.potjerodekool.codegen.template.TemplateBasedGenerator
import io.github.potjerodekool.codegen.template.model.TCompilationUnit
import io.github.potjerodekool.codegen.template.model.annotation.Annot
import io.github.potjerodekool.codegen.template.model.element.TypeElem
import io.github.potjerodekool.codegen.template.model.element.VariableElem
import io.github.potjerodekool.codegen.template.model.expression.SimpleLiteralExpr
import io.github.potjerodekool.codegen.template.model.type.ClassOrInterfaceTypeExpr
import io.github.potjerodekool.codegen.template.model.type.TypeExpr
import io.github.potjerodekool.openapi.common.ApiConfiguration
import io.github.potjerodekool.openapi.common.GeneratorConfig
import io.github.potjerodekool.openapi.common.util.OpenApiUtils.findJsonMediaType
import io.github.potjerodekool.openapi.common.util.OpenApiUtils.isImageOrVideo
import io.github.potjerodekool.openapi.common.util.OpenApiUtils.isMultiPart
import io.github.potjerodekool.openapi.common.util.SchemaWithInfo
import io.swagger.models.HttpMethod
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.Operation
import io.swagger.v3.oas.models.media.ArraySchema
import io.swagger.v3.oas.models.media.ObjectSchema
import io.swagger.v3.oas.models.media.Schema
import io.swagger.v3.oas.models.parameters.Parameter
import io.swagger.v3.oas.models.parameters.RequestBody
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.function.Consumer
import java.util.stream.Collectors
import kotlin.String

abstract class AbstractGenerator protected constructor(
    generatorConfig: GeneratorConfig,
    private val apiConfiguration: ApiConfiguration,
    val typeUtils: OpenApiTypeUtils, private val environment: Environment
) : OpenApiWalkerListener {
    val basePackageName: String = Objects.requireNonNull(
        apiConfiguration.basePackageName,
        generatorConfig::basePackageName
    )

    val modelPackageName: String = apiConfiguration.modelPackageName()

    private val language = generatorConfig.language

    private val compilationUnitMap: MutableMap<String, TCompilationUnit> = HashMap()

    fun generate(openAPI: OpenAPI) {
        OpenApiWalker(openAPI, this.apiConfiguration).walk(this)

        val generator = TemplateBasedGenerator()
        val filer = environment.filer

        val importOrganiser = ImportOrganiser()

        compilationUnitMap.values.forEach(Consumer { cu: TCompilationUnit ->
            //TODO fix importer importOrganiser.organiseImports(cu);
            val code = generator.doGenerate(cu)
            val clazz = cu.elements.first()

            val resource = filer.createResource(
                Location.SOURCE_OUTPUT,
                cu.packageName, clazz.simpleName + ".java"
            )
            resource.writeToOutputStream(code.toByteArray())
        })
    }

    protected fun findOrCreateClass(
        path: String,
        operation: Operation
    ): TypeElem {
        val className = resolveClassName(path, operation)
        val qualifiedName = QualifiedName.from(className)
        val packageName = qualifiedName.packageName().toString()
        val simpleName = qualifiedName.simpleName().toString()

        val cu = compilationUnitMap.computeIfAbsent(className) { key: String? ->
            createCompilationUnit(
                packageName,
                simpleName
            )
        }

        return cu.elements.stream()
            .findFirst()
            .orElse(null)
    }

    private fun createCompilationUnit(
        packageName: String,
        simpleName: String
    ): TCompilationUnit {
        val cu = TCompilationUnit(language)
        cu.packageName(packageName)

        val clazz = createClass(simpleName)
        cu.element(clazz)
        return cu
    }

    protected open fun createClass(simpleName: String?): TypeElem {
        val date = DateTimeFormatter.ISO_DATE_TIME.format(LocalDateTime.now())

        return TypeElem()
            .modifier(Modifier.PUBLIC)
            .simpleName(simpleName)
            .annotation(
                Annot("javax.annotation.processing.Generated")
                    .attribute(SimpleLiteralExpr(javaClass.name))
                    .attribute("date", SimpleLiteralExpr(date))
            )
    }

    protected fun resolveClassName(
        path: String,
        operation: Operation
    ): String {
        return generateClasName(path, operation)
    }

    protected fun generateClasName(
        path: String,
        operation: Operation
    ): String {
        val className = resolveQualifiedName(path, operation)
        val suffix = classNameSuffix()
        return if (suffix == null) className else className + suffix
    }

    protected abstract fun classNameSuffix(): String?

    private fun resolveQualifiedName(
        path: String,
        operation: Operation
    ): String {
        var simpleName: String?

        val pathElements = path.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        val simpleNameBuilder = StringBuilder()

        for (pathElement in pathElements) {
            if (!pathElement.startsWith("{")) {
                simpleNameBuilder.append(StringUtils.firstUpper(pathElement))
            }
        }

        simpleName = simpleNameBuilder.toString()

        simpleName = io.github.potjerodekool.openapi.common.util.StringUtils.toValidClassName(simpleName)

        val apiName: String

        if (operation.tags == null || operation.tags.isEmpty()) {
            apiName = StringUtils.firstUpper(simpleName)
        } else {
            val firstTag = operation.tags.first()
            apiName = Arrays.stream(firstTag.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
                .map { value: String? -> StringUtils.firstUpper(value) }
                .collect(Collectors.joining())
        }

        return "$basePackageName.$apiName"
    }

    protected open fun createParameters(
        api: OpenAPI,
        operation: Operation,
        httpMethod: HttpMethod
    ): MutableList<VariableElem> {
        val parameters: MutableList<VariableElem> = ArrayList()

        if (operation.parameters != null) {
            parameters.addAll(
                operation.parameters.stream()
                    .map { parameter: Parameter -> createParameter(api, parameter) }
                    .toList()
            )
        }

        parameters.addAll(createRequestBodyParameter(api, operation, httpMethod))
        return parameters
    }

    private fun createRequestBodyParameter(
        api: OpenAPI?,
        operation: Operation,
        httpMethod: HttpMethod
    ): List<VariableElem> {
        val parameters: MutableList<VariableElem> = ArrayList()
        val requestBody = operation.requestBody

        if (requestBody != null) {
            val bodyMediaType = findJsonMediaType(requestBody.content)

            if (bodyMediaType != null) {
                val resolved = SchemaResolver.resolve(api, bodyMediaType.schema)
                if (httpMethod == HttpMethod.PATCH) {
                    val bodyType = createPatchRequestBodyParameter(
                        api,
                        operation,
                        resolved,
                        bodyMediaType
                    )
                    parameters.add(createBodyParameter(operation, bodyType))
                } else {
                    if (isMultiPart(requestBody.content)) {
                        parameters.addAll(createMultiPartBodyParameters(api, requestBody))
                    } else {
                        val bodyType = createRequestBodyParameterDefault(
                            api,
                            operation,
                            bodyMediaType
                        )
                        parameters.add(createBodyParameter(operation, bodyType))
                    }
                }
            } else if (isMultiPart(requestBody.content)) {
                parameters.addAll(createMultiPartBodyParameters(api, requestBody))
            } else if (isImageOrVideo(requestBody.content)) {
                parameters.add(createImageOrVideoParameter(api))
            } else {
                val bodyType: TypeExpr = ClassOrInterfaceTypeExpr("java.lang.Object")
                parameters.add(createBodyParameter(operation, bodyType))
            }
        }

        return parameters
    }

    private fun createMultiPartBodyParameters(
        api: OpenAPI?,
        requestBody: RequestBody
    ): List<VariableElem> {
        val parameters = ArrayList<VariableElem>()

        val objectSchema = requestBody.content[ContentType.MULTIPART_FORM_DATA.type]!!.schema as ObjectSchema

        val properties = objectSchema.properties

        properties.forEach { (name: String, propertySchema: Schema<*>?) ->
            val type = typeUtils.createType(
                api,
                propertySchema,
                mapOf(),
                modelPackageName,
                null,
                null
            )
            parameters.add(createParameter(type, name))
        }
        return parameters
    }

    private fun createParameter(
        typeExpr: TypeExpr,
        name: String
    ): VariableElem {
        return VariableElem()
            .kind(ElementKind.PARAMETER)
            .modifier(Modifier.FINAL)
            .type(typeExpr)
            .simpleName(name)
    }

    private fun createImageOrVideoParameter(api: OpenAPI?): VariableElem {
        val bodyType = typeUtils.createMultipartTypeExpression(api)
        return createParameter(bodyType, "body")
    }

    protected open fun createBodyParameter(operation: Operation?, bodyType: TypeExpr): VariableElem {
        return createParameter(bodyType, "body")
    }

    private fun createRequestBodyParameterDefault(
        api: OpenAPI?,
        operation: Operation,
        bodyMediaType: SchemaWithInfo
    ): TypeExpr {
        val requestBody = operation.requestBody

        return typeUtils.createType(
            api,
            bodyMediaType.schema,
            null,
            modelPackageName,
            bodyMediaType.mediaType,
            requestBody.required
        )
    }


    private fun createPatchRequestBodyParameter(
        api: OpenAPI?,
        operation: Operation,
        resolved: ResolvedSchemaResult,
        bodyMediaType: SchemaWithInfo
    ): TypeExpr {
        val requestBody = operation.requestBody
        val bodyType: TypeExpr

        var schemaName = resolved.name
        val bodyObjectSchema = resolved.schema

        val type = typeUtils.createType(
            api,
            bodyObjectSchema,
            null,
            modelPackageName,
            bodyMediaType.mediaType,
            requestBody.required
        ) as ClassOrInterfaceTypeExpr
        bodyType = type

        if (bodyObjectSchema !is ArraySchema) {
            var patchType = type

            if (patchType.typeArguments != null && patchType.typeArguments.size == 1) {
                val resolvedItem = SchemaResolver.resolve(api, resolved.schema.items)
                schemaName = resolvedItem.name
                patchType = type.typeArguments.first() as ClassOrInterfaceTypeExpr
            }

            val simpleName = if (!schemaName.lowercase(Locale.getDefault()).contains("patch")
            ) "Patch$schemaName"
            else schemaName
            patchType.name("$modelPackageName.$simpleName")
        }

        return bodyType
    }

    protected open fun createParameter(
        openAPI: OpenAPI,
        openApiParameter: Parameter
    ): VariableElem {
        var type = typeUtils.createType(
            openAPI,
            openApiParameter.schema,
            null,
            modelPackageName,
            null,
            openApiParameter.required
        )

        if (openApiParameter.required == true) {
            type = typeUtils.asNonNull(type)
        }

        return VariableElem()
            .kind(ElementKind.PARAMETER)
            .modifier(Modifier.FINAL)
            .type(type)
            .simpleName(openApiParameter.name)
    }
}
